src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -5,73 +5,223 @@
+
+                cachedVersionStamp = tuple.Item1;
+                previousResults = tuple.Item2;
+            }
+
+            if (documentVersion == null || cachedVersionStamp != documentVersion)
+            {
+                var razorSemanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument);
+                var csharpSemanticRanges = await GetCSharpSemanticRangesAsync(codeDocument, textDocumentIdentifier, range: null, cancellationToken);
+
+                var combinedSemanticRanges = CombineSemanticRanges(razorSemanticRanges, csharpSemanticRanges);
+
+                var newTokens = ConvertSemanticRangesToSemanticTokens(combinedSemanticRanges, codeDocument, documentVersion);
+
+                if (previousResults is null)
+                {
+                    return newTokens;
+                }
+
+                var razorSemanticEdits = SemanticTokensEditsDiffer.ComputeSemanticTokensEdits(newTokens, previousResults);
+
+                return razorSemanticEdits;
+            }
+            else
+            {
+                var result = new SemanticTokensFullOrDelta(new SemanticTokensDelta
+                {
+                    Edits = Array.Empty<SemanticTokensEdit>(),
+                    ResultId = previousResultId,
+                });
+
+                return result;
+            }
+        }
+
+        private async Task<RazorCodeDocument> GetRazorCodeDocument(DocumentSnapshot documentSnapshot)
+        {
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
 
-            if (previousResultId != null)
+            return codeDocument;
+        }
+
+        private IReadOnlyList<SemanticRange> CombineSemanticRanges(params IReadOnlyList<SemanticRange>[] rangesArray)
+        {
+            var newList = new List<SemanticRange>();
+            foreach (var list in rangesArray)
+            {
+                if (list != null)
+                {
+                    newList.AddRange(list);
+                }
+            }
+
+            // Because SemanticToken data is generated relative to the previous token it must be in order.
+            // We have a guarentee of order within any given language server, but the interweaving of them can be quite complex.
+            // Rather than attempting to reason about transition zones we can simply order our ranges since we know there can be no overlapping range.
+            newList.Sort();
+
+            return newList;
+        }
+
+        private async Task<IReadOnlyList<SemanticRange>> GetCSharpSemanticRangesAsync(
+            RazorCodeDocument codeDocument,
+            TextDocumentIdentifier textDocumentIdentifier,
+            Range range,
+            CancellationToken cancellationToken)
+        {
+            var csharpDocument = codeDocument.GetCSharpDocument();
+
+            var parameter = new SemanticTokensParams
```

Comment:
In the 16.9 timeframe we're going to want to not be asking C# for the entire documents semantic tokens list. Large documents will make this whole thing fall over pretty quickly. No action for now but definitely want you to start investing some thought into this area as I imagine it'll come up real quick from users with bigger documents.